### PR TITLE
Add SQL specific tests for messages, part 2

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Rewrite queue metadata APIs similar to transfer tasks, timer tasks
* Add queue metadata tests
* Add missing encoding type

**Why?**
This PR provides better test coverage and code management.

**How did you test it?**
Run tests.

**Potential risks**
Minor business logic code change.

### CODE REVIEW

**Comment by wxing1292** [comment]
Probably need to land this PR along with proper refactoring of Cassandra counterpart

File: common/persistence/serialization/blob.go [hunk 1]
```diff
@@ -34,15 +34,16 @@ import (
...
 	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/codec"
 )
 
 func ShardInfoToBlob(info *persistencespb.ShardInfo) (commonpb.DataBlob, error) {
 	return proto3Encode(info)
 }
 
-func ShardInfoFromBlob(b []byte, proto string, clusterName string) (*persistencespb.ShardInfo, error) {
+func ShardInfoFromBlob(binary []byte, encoding string, clusterName string) (*persistencespb.ShardInfo, error) {
```
Comment (alexshtin) [inline common/persistence/serialization/blob.go:44]: I believe `b` stands for `blob` not `binary`.

File: common/persistence/sql/sqlplugin/tests/mysql_test.go [hunk 2]
```diff
@@ -90,6 +90,23 @@ func TestMySQLQueueMessageSuite(t *testing.T) {
 	suite.Run(t, s)
 }
 
+func TestMySQLQueueMetadataSuite(t *testing.T) {
+	cfg := NewMySQLConfig()
+	SetupMySQLDatabase(cfg)
+	SetupMySQLSchema(cfg)
+	store, err := sql.NewSQLDB(cfg)
+	if err != nil {
+		panic(fmt.Sprintf("unable to create MySQL DB: %v", err))
```
Comment (alexshtin) [inline common/persistence/sql/sqlplugin/tests/mysql_test.go:99]: Why not `t.Fatal`? I don't like `panic` in tests.

File: proto/internal/temporal/server/api/persistence/v1/queue.proto [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+syntax = "proto3";
+
+package temporal.server.api.persistence.v1;
+option go_package = "go.temporal.io/server/api/persistence/v1;persistence";
+
+import "google/protobuf/duration.proto";
+import "google/protobuf/timestamp.proto";
+
+// queue column
```
Comment (alexshtin) [inline proto/internal/temporal/server/api/persistence/v1/queue.proto:29]:
```suggestion
// message_payload column
```

File: proto/internal/temporal/server/api/persistence/v1/queue.proto [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+message QueueMessage {
+
+}
+
+// queue metadata column
```
Comment (alexshtin) [inline proto/internal/temporal/server/api/persistence/v1/queue.proto:34]:
```suggestion
// data column
```
And it seems this should go to `queue_metadata.proto` file.

File: common/persistence/serialization/blob.go [hunk 5]
```diff
@@ -223,16 +224,83 @@ func ChecksumToBlob(checksum *persistencespb.Checksum) (commonpb.DataBlob, error
...
+) (commonpb.DataBlob, error) {
+	if object == nil {
+		return commonpb.DataBlob{
+			Data:         nil,
+			EncodingType: encoding,
+		}, nil
+	}
+
+	switch encoding {
+	case enumspb.ENCODING_TYPE_JSON:
```
Comment (alexshtin) [inline common/persistence/serialization/blob.go:269]: I am surprised to see JSON here.

File: common/persistence/serialization/blob.go [hunk 6]
```diff
@@ -223,16 +224,83 @@ func ChecksumToBlob(checksum *persistencespb.Checksum) (commonpb.DataBlob, error
...
+	return encode(metadata, enumspb.ENCODING_TYPE_JSON)
+}
+
+func QueueMetadataFromBlob(binary []byte, encoding string) (*persistencespb.QueueMetadata, error) {
+	result := &persistencespb.QueueMetadata{}
+	return result, decode(binary, encoding, result)
+}
+
+func encode(
+	object interface{},
```
Comment (alexshtin) [inline common/persistence/serialization/blob.go:258]:
```suggestion
	object proto.Message,
```

and change `proto3Encode` to accept `proto.Message` also, and then change:
```
data, err := m.Marshal()
```
to 
```
	data, err := proto.Marshal(m)
```
there.

File: common/persistence/serialization/blob.go [hunk 7]
```diff
@@ -223,16 +224,83 @@ func ChecksumToBlob(checksum *persistencespb.Checksum) (commonpb.DataBlob, error
...
+		return nil
+	}
+
+	switch enumspb.EncodingType(enumspb.EncodingType_value[encoding]) {
+	case enumspb.ENCODING_TYPE_JSON:
+		return codec.NewJSONPBEncoder().Decode(binary, result.(proto.Message))
+	case enumspb.ENCODING_TYPE_PROTO3:
+		return proto3Decode(binary, encoding, result.(proto.Unmarshaler))
+	default:
+		return fmt.Errorf("unknown decoding type: %v", encoding)
```
Comment (alexshtin) [inline common/persistence/serialization/blob.go:300]:
```suggestion
		return fmt.Errorf("unknown encoding type: %v", encoding)
```

File: common/persistence/serialization/blob.go [hunk 8]
```diff
@@ -223,16 +224,83 @@ func ChecksumToBlob(checksum *persistencespb.Checksum) (commonpb.DataBlob, error
...
 
+func QueueMessageToBlob(message *persistencespb.QueueMessage) (commonpb.DataBlob, error) {
+	// TODO change ENCODING_TYPE_JSON to ENCODING_TYPE_PROTO3
+	return encode(message, enumspb.ENCODING_TYPE_JSON)
+}
+
+func QueueMessageFromBlob(binary []byte, encoding string) (*persistencespb.QueueMessage, error) {
+	result := &persistencespb.QueueMessage{}
+	return result, decode(binary, encoding, result)
+}
```
Comment (alexshtin) [inline common/persistence/serialization/blob.go:245]: There are not called. Where do you plan to use it?

Comment (wxing1292) [reply common/persistence/serialization/blob.go]: there will be a part 3, message itself is missing encoding

File: common/persistence/serialization/blob.go (hunk 5 above)
Comment (wxing1292) [reply common/persistence/serialization/blob.go:269]: me too: 
https://github.com/temporalio/temporal/blob/b70271b98d49927953c6916a56ceac573002916c/common/persistence/sql/sqlplugin/mysql/queue.go#L93

**Review by alexshtin — APPROVED** [review]

**Pull request merged**